### PR TITLE
Routed gather for sharded index_select and integer-tensor indexing

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

Replaces the all-gathering `ShardedIndexSelect`
with a routed exchange: a `Shard(d)` source and an index of global positions
(sharded or replicated) move only the requested rows between ranks. This is
what lets the domain-parallel mesh readers (next PR) keep global vertex ids
in the cell arrays and resolve `points[cells]` without materializing the
full point array on any rank.

There is a lot of intellectual overlap here with the halo work that alchemi
supports.  Therefore, In this PR a new primitive `exchange.py` (like `ring.py`
and `halo.py` is introduced to support both, and used as a shared backend.
The Alchemi pieces shouldn't be affected by this...

- `shard_utils/exchange.py` (new): row all-to-all, group-name resolution and
  the fp32 scatter-accumulator policy (float64 default, configurable),
  shared by `halo_scatter` and `index_ops`.
- `index_ops.py`: `routed_gather` / `sharded_index_select`; handlers for
  `torch.index_select`, `Tensor.index_select` and `Tensor.__getitem__` with a
  single int-tensor key. All exchange sizes come from the specs, so there is
  no device-to-host sync and the ops trace under `torch.compile`. The
  backward reuses the forward's request exchange and is a single collective;
  a replicated index splits backward responsibility across ranks. Negative
  indices wrap; 1-D mesh and integer dtype are enforced.
- `_ToTorchTensor.backward` adopts the gradient's shard order with its
  placements.
- Tests: `test_index.py` (getitem), `index_select` along the sharded dim in
  `test_select.py`, `test_exchange.py`, routed-gather compile test.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

---

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: physicsnemo/domain_parallel/shard_utils/exchange.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+import torch.distributed as dist
+import torch.distributed._functional_collectives as funcol
+from torch.distributed.device_mesh import DeviceMesh
+
+__all__ = [
+    "funcol_all_to_all_v_rows",
+    "get_fp32_scatter_accumulator",
+    "resolve_group_name",
+    "set_fp32_scatter_accumulator",
+]
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/exchange.py:39]: This file, `exchange.py`, is a brand new file that consolidates the halo exchange work supporting alchemi merged in v2.2.  It was useful and necessary to reproduce almost all of that for `index` and `index_select` handlers properly, so I moved core functionality here to mirror `halo.py` and `ring.py` as a slightly-lower-level utility than ShardTensor.  We could do optimizations on these functions and it would now hit all paths, etc.

File: physicsnemo/domain_parallel/shard_utils/halo_scatter.py [hunk 2]
```diff
@@ -46,62 +54,6 @@
 ]
 
 
-def _funcol_group_arg(group: object) -> object:
-    r"""Return *group* in the form functional collectives accept (funcol rejects ``None``)."""
-    if isinstance(group, DeviceMesh):
-        return (group, 0)
-    if group is None:
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/halo_scatter.py:53]: The changes to this file should be _exclusively_ moving things to `exchange.py` without functionality change.

File: physicsnemo/domain_parallel/shard_utils/index_ops.py [hunk 3]
```diff
@@ -30,214 +33,686 @@
...
+# (``request_sizes``, from the index's sharding shapes). So the exchange uses
+# STATIC split sizes and no host sync:
+#
+#   1. Every rank sends its full request list to every owner (sizes:
+#      ``request_sizes``).
+#   2. Each owner gathers, for each requester, a full-length row buffer in
+#      request order -- rows it does not own are filler -- and sends it back.
+#   3. The requester holds ``W`` buffers of its own length and picks, per
+#      request, the row from the owner it computed locally (``bucketize``).
+#
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/index_ops.py:69]: Major changes here; we never properly had an algorithm here, it was always a hack.  Now, it's sort of an algorithm.  RDMA could help but it's not really part of physics nemo or anything yet.

File: physicsnemo/domain_parallel/shard_utils/index_ops.py [hunk 4]
```diff
@@ -30,214 +33,686 @@
...
+def _wrap_negative(index_flat: torch.Tensor, n_global: int) -> torch.Tensor:
+    r"""Wrap negative global positions like eager indexing does.
+
+    Parameters
+    ----------
+    index_flat : torch.Tensor
+        1-D global positions; negatives count from the end.
+    n_global : int
+        Global extent along the indexed dim. Positions ``>= n_global`` are not
+        checked (they select filler rows on the owner side).
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/index_ops.py:99]: This is a handler for negative indexes, which is good practice for sanitizing our arguments here.

File: physicsnemo/domain_parallel/shard_utils/index_ops.py [hunk 5]
```diff
@@ -30,214 +33,686 @@
...
-        ------
-        MissingShardPatch
-            If the index sharding strategy is not implemented.
-        """
-        # This is the simplest implementation, to enable functionality.
-        # It could be optimized for very large tensors to ensure performace.
-
-        # First - Make sure we have the full input tensor
-        # Triggers an all_gather(_v) for (uneven) tensors.
-        local_tensor = tensor.full_tensor()
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/index_ops.py:83]: This was bad news previously.

File: physicsnemo/domain_parallel/shard_utils/index_ops.py [hunk 6]
```diff
@@ -30,214 +33,686 @@
...
-
-            - Gradient with respect to input tensor
-            - ``None`` for dim parameter (not differentiable)
-            - ``None`` for index parameter (not differentiable)
-        """
-        (index,) = ctx.saved_tensors
-        spec = ctx.spec
-        dim = ctx.dim
-
-        local_index = index.full_tensor()
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/index_ops.py:190]: More bad news.

File: physicsnemo/domain_parallel/shard_utils/index_ops.py [hunk 7]
```diff
@@ -30,214 +33,686 @@
...
-
-        grad_inputs = torch.zeros(
-            spec.tensor_meta.shape,
-            device=grad_output._local_tensor.device,
-            dtype=grad_output._local_tensor.dtype,
-        )
-        # local_grad_output = grad_output.to_local()
-        local_grad_output = grad_output.full_tensor()
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/index_ops.py:198]: still more bad news.

File: physicsnemo/domain_parallel/shard_utils/index_ops.py [hunk 8]
```diff
@@ -30,214 +33,686 @@
...
+# ``values`` (``offsets``) and how many requests every rank holds
+# (``request_sizes``, from the index's sharding shapes). So the exchange uses
+# STATIC split sizes and no host sync:
+#
+#   1. Every rank sends its full request list to every owner (sizes:
+#      ``request_sizes``).
+#   2. Each owner gathers, for each requester, a full-length row buffer in
+#      request order -- rows it does not own are filler -- and sends it back.
+#   3. The requester holds ``W`` buffers of its own length and picks, per
+#      request, the row from the owner it computed locally (``bucketize``).
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/index_ops.py:68]: The algorithm here *could* be a two-stage pass of 
- ask everyone, "what do you need, on all ranks?" in an all-gather
- then, it's an all-to-allv to collect that.

The reason I didn't do that, even though it's less bandwidth, is that you would necessarily have a cuda.sync to allocate ragged memory buffers based on the size of arrays from each other rank.  Instead, I take a path where we over-allocate and over-spend on comms, since it's still relatively cheap, and have 100% non-blocking ops.

File: physicsnemo/domain_parallel/shard_utils/index_ops.py [hunk 9]
```diff
@@ -30,214 +33,686 @@
...
+        return routed_gather(tensor, index, dim)
+
+    if placement.is_shard():
+        if index_sharded:
+            if _is_tracing((tensor, index)):
+                raise MissingShardPatch(
+                    "index_select off the sharded dim with a sharded index needs an "
+                    "all-gather of the index; not supported under torch.compile"
+                )
+            idx = index.full_tensor()
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/index_ops.py:692]: In this case, there is not a nice way I've come up with to avoid resolving the full index tensor here.

File: physicsnemo/domain_parallel/shard_tensor.py [hunk 10]
```diff
@@ -183,10 +183,14 @@ def backward(ctx, grad_output: DTensor):
...
         # and stamping the primal's stride onto differently-laid-out grad
-        # memory breaks downstream .view() calls.
+        # memory breaks downstream .view() calls. The shard order must follow
+        # the placements: DTensorSpec asserts that every Shard placement has a
+        # matching entry, so a Shard primal with a Replicate cotangent (a
+        # replicated result gathered from sharded rows) would otherwise fail.
         cached_spec = dataclasses.replace(
             cached_spec,
             placements=grad_placements,
+            shard_order=grad_output._spec.shard_order,
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_tensor.py:193]: shard_order is an upstream attribute we're simply adding in here.
